File: src/348. Sum of a square and a cube/348.py
import math

def is_palindrome(n):
    return str(n) == str(n)[::-1]

# if the result is not enough, take a bigger guess
upper_limit = 10 ** 9

squares = []
for a in range(2, int(math.sqrt(upper_limit))):
    squares.append(a * a)

cubes = []
b = 2
while b * b * b < upper_limit:
    cubes.append(b * b * b)
    b += 1

candidates = {}
for s in squares:
    for c in cubes:
        if is_palindrome(s + c):
            count = candidates.get(s + c, 0)
            candidates[s + c] = count + 1

result = []
for k, v in candidates.items():
    if v == 4:
        result.append(k)

if len(result) >= 5:
    result.sort()
    print(sum(result[0:5]))

File: src/348. Sum of a square and a cube/348.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Main {

    private static final long UPPER_LIMIT = 1_000_000_000L;

    public static void main(String[] args) {
        // if the result is not enough, take a bigger guess

        Set<Long> squares = new HashSet<>();
        for (long a = 2; a < Math.sqrt(UPPER_LIMIT); a++) {
            squares.add(a * a);
        }

        Set<Long> cubes = new HashSet<>();
        for (long b = 2; b * b * b < UPPER_LIMIT; b++) {
            cubes.add(b * b * b);
        }

        Map<Long, Integer> candidates = new HashMap<>();
        for (long s : squares) {
            for (long c : cubes) {
                if (isPalindrome(s + c)) {
                    int count = candidates.getOrDefault(s + c, 0);
                    candidates.put(s + c, count + 1);
                }
            }
        }

        List<Long> result = new ArrayList<>();
        for (Map.Entry<Long, Integer> kv : candidates.entrySet()) {
            if (kv.getValue() == 4) {
                result.add(kv.getKey());
            }
        }

        if (result.size() >= 5) {
            result.sort(Long::compareTo);
            long sum = result.stream()
                    .mapToLong(Long::longValue)
                    .limit(5)
                    .sum();
            System.out.println(sum);
        }
    }

    private static boolean isPalindrome(long n) {
        String original = Long.toString(n);
        return reverseString(original).equals(original);
    }

    private static String reverseString(String original) {
        return new StringBuilder(original).reverse().toString();
    }

}
